File: Part 2/csharp_solution/Exchange/OrderBook.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Medallion.Collections;

namespace Exchange
{
    class BuyComparer : IComparer<Order>
    {
        public int Compare(Order left, Order right)
        {
            if (left.Price == right.Price) {
                return left.Generation.CompareTo(right.Generation);
            }

            return right.Price.CompareTo(left.Price);
        }
    }

    class SellComparer : IComparer<Order>
    {
        public int Compare(Order left, Order right)
        {
            if (left.Price == right.Price) {
                return left.Generation.CompareTo(right.Generation);
            }

            return left.Price.CompareTo(right.Price);
        }
    }

    public class OrderBook
    {
        public Trade[] Execute(Order order)
        {
            if (order.Quantity > 0) {
                BuyOrders.Enqueue(order);
            }    
            else {
                SellOrders.Enqueue(order);
            }

            var trades = new List<Trade>();

            while (BuyOrders.Count > 0 && SellOrders.Count > 0) {

                var buy = BuyOrders.Peek();
                var sell = SellOrders.Peek();

                if (buy.Price < sell.Price) {
                    break;
                }

                var quantity = Math.Min((buy.RemainingQuantity), sell.RemainingQuantity);
                var price = buy.Generation < sell.Generation ? buy.Price : sell.Price;

                var trade = new Trade(buy.Participant, sell.Participant, buy.Instrument, quantity, price);

                trades.Add(trade);

                buy.Fill(quantity);
                sell.Fill(quantity);

                if (buy.RemainingQuantity == 0) {
                    BuyOrders.Dequeue();
                }

                if (sell.RemainingQuantity == 0) {
                    SellOrders.Dequeue();
                }
            }
            
            return trades.ToArray();
        }

        public PriorityQueue<Order> BuyOrders = new PriorityQueue<Order>(new BuyComparer());
        public PriorityQueue<Order> SellOrders = new PriorityQueue<Order>(new SellComparer());

    }
}

File: Part 2/java_solution/src/com/geh/Exchange/OrderBook.java
package com.geh.Exchange;

import java.util.PriorityQueue;
import java.util.LinkedList;

public class OrderBook {

    public Trade[] execute(Order order)
    {
        if (order.getQuantity() > 0) {
            buyOrders.add(order);
        }
        else {
            sellOrders.add(order);
        }

        var trades = new LinkedList<Trade>();

        while (!buyOrders.isEmpty() && !sellOrders.isEmpty()) {

            var buyOrder = buyOrders.peek();
            var sellOrder = sellOrders.peek();

            if (buyOrder.getPrice().compareTo(sellOrder.getPrice()) < 0) {
                break;
            }

            var quantity = Math.min(buyOrder.getRemainingQuantity(), sellOrder.getRemainingQuantity());
            var price = buyOrder.getGeneration() < sellOrder.getGeneration() ? buyOrder.getPrice() : sellOrder.getPrice();

            var trade = new Trade(buyOrder.getParticipant(), sellOrder.getParticipant(), buyOrder.getInstrument(), quantity, price);

            trades.add(trade);

            buyOrder.fill(quantity);
            sellOrder.fill(quantity);

            if (buyOrder.getRemainingQuantity() == 0) {
                buyOrders.remove(buyOrder);
            }

            if (sellOrder.getRemainingQuantity() == 0) {
                sellOrders.remove(sellOrder);
            }
        }

        return trades.toArray(Trade[]::new);
    }

    public PriorityQueue<Order> getBuyOrders() {
        return buyOrders;
    }

    public PriorityQueue<Order> getSellOrders() {
        return sellOrders;
    }

    private PriorityQueue<Order> buyOrders = new PriorityQueue<>(1000, (left, right) -> {
        var comparison = right.getPrice().compareTo(left.getPrice());
        if (comparison == 0) {
            return Integer.compare(left.getGeneration(), right.getGeneration());
        }
        return comparison;
    });

    private PriorityQueue<Order> sellOrders = new PriorityQueue<>(1000, (left, right) -> {
        var comparison = left.getPrice().compareTo(right.getPrice());
        if (comparison == 0) {
            return Integer.compare(left.getGeneration(), right.getGeneration());
        }
        return comparison;
    });

}
